v4/internal/integrationsupport/integrationsupport.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package integrationsupport exists to expose functionality to integration
+// packages without adding noise to the public API.
+package integrationsupport
+
+import (
+	"github.com/newrelic/go-agent/v4/internal"
+	"github.com/newrelic/go-agent/v4/newrelic"
+	"go.opentelemetry.io/otel/api/trace/testtrace"
+)
+
+// AddAgentAttribute allows instrumentation packages to add agent attributes.
+func AddAgentAttribute(txn *newrelic.Transaction, id string, stringVal string, otherVal interface{}) {
+}
+
+// AddAgentSpanAttribute allows instrumentation packages to add span attributes.
+func AddAgentSpanAttribute(txn *newrelic.Transaction, key string, val string) {}
+
+// SampleAppName is a sample application name.
+const SampleAppName = "my app"
+
+// ExpectApp combines Application and Expect, for use in validating data in test apps
+type ExpectApp struct {
+	internal.Expect
+	*newrelic.Application
+}
+
+// ConfigFullTraces enables distributed tracing and sets transaction
+// trace and transaction trace segment thresholds to zero for full traces.
+func ConfigFullTraces(cfg *newrelic.Config) {
+	cfg.DistributedTracer.Enabled = true
+	cfg.TransactionTracer.Segments.Threshold = 0
+	cfg.TransactionTracer.Threshold.IsApdexFailing = false
+	cfg.TransactionTracer.Threshold.Duration = 0
+}
+
+// NewTestApp creates an ExpectApp with the given ConnectReply function and Config function
+func NewTestApp(cfgFn ...newrelic.ConfigOption) ExpectApp {
+	tr := testtrace.NewProvider().Tracer("go-agent-test")
+	cfgFn = append(cfgFn, func(cfg *newrelic.Config) {
+		cfg.OpenTelemetry.Tracer = tr
+	})
+
+	app, err := newrelic.NewApplication(cfgFn...)
+	if nil != err {
+		panic(err)
+	}
+
+	return ExpectApp{
+		Expect: &internal.TraceExpecter{
+			Tracer: tr.(*testtrace.Tracer),
+		},
```

Comment:
This is the most important and complex change in this PR. Ultimately, we want our current integrations tests to still be able to run with very minimal code changes. These tests all involve calling `integrationsupport.NewTestApp` then calling the expect methods on the returned app. For example `app.ExpectSpanEvents` or `app.ExpectMetrics`.

In the v3 incarnation of the agent, the implementation of these `app.Expect*` methods were in the `newrelic` package proper and were compiled into agent binaries. (Maybe they were compiled out by the compiler, but either way, they lived in the agent API code proper) I've always felt this was a mistake. So, instead I've chosen to put that code in `internal` instead as part of a new `internal.TraceExpecter` type. This way it will only ever be used when we are specifically testing.